### PR TITLE
feat: finish LedgerId

### PR DESCRIPTION
**Pull request by Sean-Tedrow-LB**
Signed-off-by: Sean Tedrow <sean.tedrow@launchbadge.com>

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [X] Tested (unit, integration, etc.)

### CODE REVIEW

File: sdk/rust/src/ledger_id.rs [hunk 1]
```diff
@@ -45,6 +55,10 @@ impl LedgerId {
     pub fn is_known_network(&self) -> bool {
         self.is_mainnet() || self.is_previewnet() || self.is_testnet()
     }
+
+    pub const fn to_bytes(&self) -> &Vec<u8> {
```
Comment (ghost) [inline sdk/rust/src/ledger_id.rs]: that would be `as_bytes` with the current signature (https://rust-lang.github.io/api-guidelines/naming.html#ad-hoc-conversions-follow-as_-to_-into_-conventions-c-conv)... and technically you'd want to return a ` &[u8]` (there's basically nothing you can do with a `&Vec<T>` that you can't do with a `&[u8]`)

anyway, for `to_bytes` you'd want to return a `Vec<u8>` (`self.0.clone()`)

File: sdk/rust/src/ledger_id.rs [hunk 2]
```diff
@@ -55,6 +69,66 @@ impl Debug for LedgerId {
...
     fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
-        f.pad(&base64::encode(&self.0))
+        if self.is_mainnet() {
+            f.write_str("mainnet")
+        } else if self.is_testnet() {
+            f.write_str("testnet")
+        } else if self.is_previewnet() {
+            f.write_str("previewnet")
+        } else {
+            f.write_str(&hex::encode(self.to_bytes()))
```
Comment (ghost) [inline sdk/rust/src/ledger_id.rs]: following the above, `to_*` is generally considered to be "not free" so the field access would be the preferred way of doing things in this particular case

File: sdk/rust/src/ledger_id.rs [hunk 3]
```diff
@@ -55,6 +69,66 @@ impl Debug for LedgerId {
...
+
+impl FromStr for LedgerId {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "mainnet" => Ok(Self::mainnet()),
+            "testnet" => Ok(Self::testnet()),
+            "previewnet" => Ok(Self::previewnet()),
+            _ => hex::decode(s).map(Self::from_bytes).map_err(Error::basic_parse),
```
Comment (ghost) [inline sdk/rust/src/ledger_id.rs:99]: do you need to strip a `0x` prefix here?

File: sdk/rust/src/ledger_id.rs [hunk 4]
```diff
@@ -26,6 +32,10 @@ impl LedgerId {
         Self(vec![2])
     }
 
+    pub fn from_bytes(bytes: Vec<u8>) -> Self {
+        Self(bytes)
+    }
+
     #[must_use]
     pub fn is_mainnet(&self) -> bool {
```
Comment (ghost) [inline sdk/rust/src/ledger_id.rs:41]: (not to this line, and maybe a future PR type thing, I can make an issue if so)
it'd be nice if there was a `pub(crate)` way of getting the network (motivation: more understandable code? Subjectively)

```rs
// don't really need `Eq` or `PartialEq`
#[derive(Copy, Clone)]
pub(crate) enum KnownNetwork {
    Mainnet,
    Testnet,
    Previewnet,
}

impl KnownNetwork {
    #[must_use]
    pub(crate) const fn as_str(self) -> &'static str {
        todo!()
    }
}

impl LedgerId {
    #[must_use]
    pub(crate) const fn network(&self) -> Option<KnownNetwork> {
        match &self.0 {
             [0] => Some(KnownNetwork::Mainnet),
             // ...
    }

    #[must_use]
    pub fn is_mainnet(&self) -> bool {
        matches!(self.network(), Some(KnownNetwork::Mainnet)
    }

    // ... other `is_*` functions ...

    #[must_use]
    pub fn is_known_network(&self) -> bool {
        self.network().is_some()
    }
}

impl Display for LedgerId {
    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        // can use `self.network()`
    }
```

**Review by ghost — CHANGES_REQUESTED** [review]
One fairly ignorable comment, one question and a related pair of requested changes

Good stuff.

File: sdk/rust/src/ledger_id.rs (hunk 1 above)
Comment (Sean-Tedrow-LB) [reply sdk/rust/src/ledger_id.rs]: `to_bytes()` is specified in the SDK spec.  Should we have both `as_bytes()` and `to_bytes()`?  Is there any reason to have the `to_bytes()` implementation you suggested here?  Couldn't the user simply `clone()` the return value themselves if they want?

I will go ahead and change the return type.

File: sdk/rust/src/ledger_id.rs (hunk 3 above)
Comment (Sean-Tedrow-LB) [reply sdk/rust/src/ledger_id.rs:99]: The java SDK does not.

File: sdk/rust/src/ledger_id.rs (hunk 4 above)
Comment (Sean-Tedrow-LB) [reply sdk/rust/src/ledger_id.rs:41]: We had something similar to this in the Java SDK called `NetworkName` (terrible name, I know), and we deprecated it in favor of `LedgerId` and it's not used anywhere outside of deprecated functions.  From my experience there, I don't think we need this.

Comment (izik1) [reply sdk/rust/src/ledger_id.rs:41]: fair, I meant as an internal only thing, to be clear, but if you don't think it'd work that's fine

**Review by ghost — APPROVED** [review]

**Pull request merged**